### PR TITLE
Added new profile/setting pages, updated design, removed UNIQUE constraints from problem and assignment titles

### PR DESCRIPTION
**Pull request by emmetuft**

### CODE REVIEW

File: front_end/content.py [hunk 1]
```diff
@@ -154,6 +167,12 @@ def create_sqlite_tables(self):
         else:
             print("Error! Cannot create a database connection.")
 
+    def add_column(self):
```
Comment (srp33) [inline front_end/content.py:170]: The name of this function suggests that it is generic to any column, but it appears to be adding one specific column. I think it should either be more generic or else the name should be more specific. Or maybe better, shouldn't this be in a migration script only rather than here?

File: front_end/content.py [hunk 2]
```diff
@@ -330,6 +379,41 @@ def get_user_count(self):
...
+        sql = '''SELECT COUNT(*) AS count
+                 FROM courses
+                 WHERE course_id = ?'''
+        self.cursor.execute(sql, (course_id,))
+        if self.cursor.fetchone():
+            return True
+        else:
+            return False
+
+    def get_course_from_role(self, user_id):
```
Comment (srp33) [inline front_end/content.py:392]: This doesn't seem to be related to roles specifically. Should you have a role in the WHERE clause? Please make the function name more specific, too.

File: front_end/content.py [hunk 3]
```diff
@@ -330,6 +379,41 @@ def get_user_count(self):
...
+                 ON p.course_id = c.course_id
+                 WHERE user_id = ?'''
+
+        self.cursor.execute(sql, (user_id,))
+        for course in self.cursor.fetchall():
+            course_basics = {"id": course["course_id"], "title": course["title"]}
+            courses.append([course["course_id"], course_basics])
+        return courses
+
+    def get_course_passcode(self, course_id):
```
Comment (srp33) [inline front_end/content.py:406]: I would rather that the passcode be returned with course details rather than having a separate function for returning just the passcode. Although just returning the passcode is a bit faster, I'm concerned about keeping the code as simple as possible.

File: front_end/content.py [hunk 4]
```diff
@@ -363,14 +447,14 @@ def get_problem_ids(self, course_id, assignment_id):
     def get_courses(self, show_hidden=True):
```
Comment (srp33) [inline front_end/content.py:447]: In this case, I would prefer to make a separate function called get_courses_details. get_courses() is used on every page in the navbar, so we want it to be simple and fast. However, in this case, we are pulling details about the course, and this is only used occasionally. For now get_courses_details can just return course_id, title, visible, and introduction. But in the future, we may want it to pull additional details.

File: front_end/content.py [hunk 5]
```diff
@@ -907,6 +1008,12 @@ def has_duplicate_title(self, entries, this_entry, proposed_title):
                 return True
         return False
 
+    def save_course_passcode(self, course_id, passcode):
```
Comment (srp33) [inline front_end/content.py:1011]: The passcode should just be saved along with the other course details when we add/edit a course, rather than in its own function.

File: front_end/content.py [hunk 6]
```diff
@@ -1014,10 +1121,17 @@ def copy_assignment(self, course_id, assignment_id, new_course_id):
...
         sql = '''UPDATE users
-                 SET user_json = ?
+                 SET name = ?, given_name = ?, family_name = ?, picture = ?, locale = ?
                  WHERE user_id = ?'''
 
-        self.cursor.execute(sql, (json.dumps(user_dict), user_id,))
+        self.cursor.execute(sql, (user_dict["name"], user_dict["given_name"], user_dict["family_name"],
+        user_dict["picture"], user_dict["locale"], user_id,))
+
+    def update_user_theme(self, user_id, theme):
```
Comment (srp33) [inline front_end/content.py:1130]: Let's make this more generic than just the theme. Maybe update_user_setting() or update_user_settings() so that in the future when we want to support additional settings, we can use the same function.

File: front_end/html/profile_header.html [hunk 7]
```diff
@@ -0,0 +1,34 @@
+<!DOCTYPE html>
```
Comment (srp33) [inline front_end/html/profile_header.html:1]: Do we need to have a separate header file? It would be preferable if we only have one so that we only have one place where we need to change things when the header needs to change.

File: front_end/webserver.py [hunk 8]
```diff
@@ -145,20 +154,234 @@ def get(self):
...
+            if role == "administrator":
+                registered_courses = content.get_courses()
+            elif role == "instructor" or role == "assistant":
+                registered_courses = content.get_course_from_role(user_id)
+            else:
+                registered_courses = content.get_registered_courses(user_id)
+            self.render("profile_courses.html", page="courses", result=None, courses=content.get_courses(), registered_courses=registered_courses, user_info=content.get_user_info(user_id), user_logged_in=user_logged_in_var.get(), role=role)
+        except Exception as inst:
+            render_error(self, traceback.format_exc()) 
+    def post(self, user_id):
```
Comment (srp33) [inline front_end/webserver.py:170]: Please add blank line between functions.

File: front_end/content.py [hunk 9]
```diff
@@ -277,10 +296,40 @@ def get_course_id_from_role(self, user_id):
...
+                 WHERE course_id = ?
+                 AND user_id = ?'''
+
+        self.cursor.execute(sql, (course_id, user_id,))
+        if self.cursor.fetchone():
+            return True
+
+        return False
+
+    def get_user_info(self, user_id):
```
Comment (srp33) [inline front_end/content.py:323]: This works, but I had something a little different in mind regarding how to store the user information in a context variable. I'll tackle that part after we merge this.

File: front_end/webserver.py [hunk 10]
```diff
@@ -145,20 +154,234 @@ def get(self):
...
+
+    def post(self, course_id, user_id):
+        try:
+            role = self.get_current_role()
+            if role == "administrator" or role == "instructor":
+                new_assistant = self.get_body_argument("new_assistant")
+                result = ""
+
+                if content.check_user_exists(new_assistant):
+                    if content.get_role(new_assistant) == "assistant":
```
Comment (srp33) [inline front_end/webserver.py:278]: Checking the role for an assistant (or an instructor) should be course specific. Please add an argument to this function that checks against the current course ID. Then update the code throughout webserver.py to deal with this properly. It'll have to deal with the fact that there is no course associated with an administrator role. I believe this is an old bug, unless I am forgetting something.

**Review by srp33 — CHANGES_REQUESTED** [review]

File: front_end/content.py (hunk 1 above)
Comment (emmetuft) [reply front_end/content.py:170]: Sorry, this function was just a temporary method so I could make changes to the database one by one. It isn't needed at all.

File: front_end/html/profile_header.html (hunk 7 above)
Comment (emmetuft) [reply front_end/html/profile_header.html:1]: The header file for the profile is different because of the spacing/design. If we want to use a single file, I'll have to update the design of every other html file. So we can do that, but it will take some time.

File: front_end/webserver.py (hunk 10 above)
Comment (emmetuft) [reply front_end/webserver.py:280]: I need some help with this one.

File: front_end/content.py (hunk 4 above)
Comment (emmetuft) [reply front_end/content.py:430]: Did I just start using it to assign "registered_courses" for admins? I'm confused on this one.

**Review by srp33 — APPROVED** [review]
